test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1

```diff
@@ -715,3 +715,209 @@ Describe 'Base type has abstract properties' -Tags "CI" {
+            $c2.CallVirtualMethod2Base() | Should -Be 'C1_VirtualMethod2'
+            $c2.CallVirtualMethod2Derived() | Should -Be 'C2_VirtualMethod2'
+        }
+    }
+
+    Context 'Derived class can access instance base class members dynamically' {
+
+        It 'can access <accessType> base fields and properties' -TestCases $testCases {
+            param($derivedType)
+            $c2 = $derivedType::new()
+            $c2.GetInstanceMemberDynamic('InstanceField') | Should -Be 'C1_InstanceField'
+            $c2.GetInstanceMemberDynamic('InstanceProperty') | Should -Be 'C1_InstanceProperty'
+            $c2.GetInstanceMemberDynamic('VirtualProperty1') | Should -Be 'C1_VirtualProperty1'
+            $c2.SetInstanceMemberDynamic('InstanceField', 'foo1') | Should -Be 'foo1'
+            $c2.SetInstanceMemberDynamic('InstanceProperty', 'foo2') | Should -Be 'foo2'
+            $c2.SetInstanceMemberDynamic('VirtualProperty1', 'foo3') | Should -Be 'foo3'
+        }
+
+        It 'can call <accessType> base methods' -TestCases $testCases {
+            param($derivedType)
+            $c2 = $derivedType::new()
+            $c2.CallInstanceMemberDynamic('InstanceMethod') | Should -Be 'C1_InstanceMethod'
+            $c2.CallInstanceMemberDynamic('VirtualMethod1') | Should -Be 'C1_VirtualMethod1'
+        }
+    }
+
+    Context 'Base class members are not accessible outside class scope' {
+
+        BeforeAll {
+            $instanceTest = {
+                $c2 = $derivedType::new()
+                { $null = $c2.InstanceField } | Should -Throw -ErrorId 'PropertyNotFoundStrict'
+                { $null = $c2.InstanceProperty } | Should -Throw -ErrorId 'PropertyNotFoundStrict'
+                { $null = $c2.VirtualProperty1 } | Should -Throw -ErrorId 'PropertyNotFoundStrict'
+                { $c2.InstanceField = 'foo' } | Should -Throw -ErrorId 'PropertyAssignmentException'
+                { $c2.InstanceProperty = 'foo' } | Should -Throw -ErrorId 'PropertyAssignmentException'
+                { $c2.VirtualProperty1 = 'foo' } | Should -Throw -ErrorId 'PropertyAssignmentException'
+                { $derivedType::new().InstanceMethod() } | Should -Throw -ErrorId 'MethodNotFound'
+                { $derivedType::new().VirtualMethod1() } | Should -Throw -ErrorId 'MethodNotFound'
+                foreach ($name in @('InstanceField', 'InstanceProperty', 'VirtualProperty1')) {
+                    { $null = $c2.$name } | Should -Throw -ErrorId 'PropertyNotFoundStrict'
+                    { $c2.$name = 'foo' } | Should -Throw -ErrorId 'PropertyAssignmentException'
+                }
+                foreach ($name in @('InstanceMethod', 'VirtualMethod1')) {
+                    { $c2.$name() } | Should -Throw -ErrorId 'MethodNotFound'
+                }
+            }
+            $c3UnrelatedType = Invoke-Expression @"
+                class C3Unrelated {
+                    [void]RunInstanceTest([type]`$derivedType) { $instanceTest }
```

Comment:
You meant to invoke the script block `$instanceTest`, right?
```suggestion
                    [void]RunInstanceTest([type]`$derivedType) { $instanceTest.Invoke() }
```

Reply:
No. I want the class method `RunInstanceTest` to have the same inner code as the script block. The actual invocation happens in the `It` block below.

If no class scope is given, the script block is invoked in script scope. Otherwise, the class is instantiated and, by calling the method, the same code runs in a class scope, but not the one needed to successfully access the protected (internal) members.

Reply:
Ah, my bad. This is within an expandable string. You are good.